znnl/analysis/loss_ntk_calculation.py

```diff
@@ -0,0 +1,241 @@
+"""
+ZnNL: A Zincwarecode package.
+
+License
+-------
+This program and the accompanying materials are made available under the terms
+of the Eclipse Public License v2.0 which accompanies this distribution, and is
+available at https://www.eclipse.org/legal/epl-v20.html
+
+SPDX-License-Identifier: EPL-2.0
+
+Copyright Contributors to the Zincwarecode Project.
+
+Contact Information
+-------------------
+email: zincwarecode@gmail.com
+github: https://github.com/zincware
+web: https://zincwarecode.com/
+
+Citation
+--------
+If you use this module please cite us with:
+
+Summary
+-------
+"""
+
+from typing import Callable
+
+import jax
+import jax.numpy as np
+import neural_tangents as nt
+
+from znnl.models.jax_model import JaxModel
+
+
+class LossNTKCalculation:
+    """
+    Class to calculate the loss NTK matrix for a given model and dataset.
+
+    The loss NTK matrix is defined as
+    .. math::
+        (\\Lambda_L)_{ij} = \\sum_\\theta\\frac{\\partial\\ell_i}{\\partial\\theta}
+        \\frac{\\partial\\ell_j}{\\partial\\theta}}.
+    Where :math:`\\ell_i` is the loss function for the single datapoint :math:`i` and
+    :math:`\\theta` are the parameters of the network.
+    """
+
+    def __init__(
+        self,
+        metric_fn: Callable,
+        model: JaxModel,
+        dataset: dict,
+    ):
+        """
+        Constructor for the loss ntk calculation class.
+
+        Parameters
+        ----------
+
+        metric_fn : Callable
+                The metric function to be used for the loss calculation.
+                !This has to be the metric, not the Loss!
+                If you put in the Loss here you won't get an error but an
+                incorrect result.
+
+        model : JaxModel
+                The model for which to calculate the loss NTK.
+
+        dataset : dict
+                The dataset for which to calculate the loss NTK.
+                The dictionary should contain the keys "inputs" and "targets".
+        """
+
+        # Set the attributes
+        self.ntk_batch_size = model.ntk_batch_size
```

Comment:
It might be better to make all of these arguments in this calculation. Especially when we later move into the new measurement system, this will all need to be self contained. Things like `store_on_device` are only pertinent to this calculator.